README.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      README.md
--
--  BRIEF
--      Important information regarding this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--
------------------------------------------------------------------------------->

# cffreference

## Summary

A simple CLI to extract citation information from a CITATION.cff.

## License

This project's license is **GPL 2** (as of June 1991).  The whole license text
can be found in `LICENSE` in the main directory of this repository.  The brief
version is as follows:

> Copyright (C) 2022 Kevin Matthes
>
> This program is free software; you can redistribute it and/or modify
> it under the terms of the GNU General Public License as published by
> the Free Software Foundation; either version 2 of the License, or
> (at your option) any later version.
>
> This program is distributed in the hope that it will be useful,
> but WITHOUT ANY WARRANTY; without even the implied warranty of
> MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
> GNU General Public License for more details.
>
> You should have received a copy of the GNU General Public License along
> with this program; if not, write to the Free Software Foundation, Inc.,
> 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

When compiling a printable version of this documentation using Pandoc, the full
license will be attached automatically to the resulting document.  This can be
invoked by calling `repository-manual.m`.

## Software Requirements

| Requirement       | Type          | Role                                  |
|:------------------|:-------------:|:--------------------------------------|
| `docs-snippets`   | repository    | documentation constants               |
| GHC               | application   | Haskell compiler                      |
| GNU Octave        | application   | execution of the provided scripts     |
| Haddock           | application   | creation of source code documentation |
| Pandoc            | application   | compilation of repository manual      |
| `texlive-full`    | package       | compilation of repository manual      |

The compilation of such an **optional** repository manual can be invoked by just
calling one of the following lines in a terminal.

```
octave repository-manual.m
octave-cli repository-manual.m
```

Both will redirect to the same application, GNU Octave, which will then create
the manual for this repository and attach the entire license to it.  The
resulting file, `repository.pdf`, will be saved in the main directory of this
repository.

## Description

A common task when working with CFF is adding another project's citation meta
data to the own `CITATION.cff`'s `references` section.  Since these files are
often very long -- due to having `references` to cite on their own or many
people who contributed, for instance -- and not sorted uniformly, it might be
a challenge to filter out all required lines by hand.  Furthermore, there can
also be a `preferred-citation` provided which could be overseen in a quite big
citation configuration file.

In order to solve these issues, `cffreference` provides the capability to
appropriately cite another project automatically.  Given an input file which is
valid according to the Schema Version 1.2.0, `cffreference` will extract the
relevant lines to cite.

Any input file will be cited by its `preferred-citation`, by default.  If this
entry should not be specified, the whole file will be processed in order to
generate a proper `references` item.  In the latter case, the input file is
assumed to specify a `software`.

By default, the results are written to `stdout`.  This behaviour is intended to
be a preview mode in order to verify that the correct repository is going to be
cited.  The extracted results can be also piped around in terminal sessions such
that other CFF tools can be instructed with `cffreference`'s output.

```
cffreference ./path/to/input/CITATION.cff
```

In order to append the extracted citation information to a certain, writable
`CITATION.cff`, its path must be given as the second argument to the application
which will then add the new reference at the end of the output file.

```
cffreference ./path/to/input/CITATION.cff ./path/to/output/CITATION.cff
```

In case no arguments or more than two are given, the application will show the
license information as well as a brief information on how to use `cffreference`
correctly.

```
cffreference
```
<!----------------------------------------------------------------------------->


CHANGELOG.md
<!------------------------------------------------------------------------------
--
-- Copyright (C) 2022 Kevin Matthes
--
-- This program is free software; you can redistribute it and/or modify
-- it under the terms of the GNU General Public License as published by
-- the Free Software Foundation; either version 2 of the License, or
-- (at your option) any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License along
-- with this program; if not, write to the Free Software Foundation, Inc.,
-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
--
----
--
--  FILE
--      CHANGELOG.md
--
--  BRIEF
--      The development history of this project.
--
--  AUTHOR
--      Kevin Matthes
--
--  COPYRIGHT
--      (C) 2022 Kevin Matthes.
--      This file is licensed GPL 2 as of June 1991.
--
--  DATE
--      2022
--
--  NOTE
--      See `LICENSE' for full license.
--      See `README.md' for project details.
--
------------------------------------------------------------------------------->

# Changelog

## [Unreleased]

### Added

* Rust refactoring of Haskell code

### Deprecated

* Haskell prototype, including its build routine

## [0.5] -- 2022-05-13

### Added

* create `.github/workflows/cffconvert.yml`
* support further additional placeholders in input files

### Changed

* adjust README description
* update `.gitignore`

## [0.4] -- 2022-05-12

### Changed

* apply the configured type of the `preferred-citation`
* optimise comment banners of `CITATION.cff`
* outsource `extract` to a dedicated function
* outsource `skipReferences` to a dedicated function
* set default citation type to `software`

### Fixed

* add missing patterns in `Main.hs`
* fix behaviour of `extract`

## [0.3] -- 2022-05-11

### Added

* cite `blank`
* cite `docs-snippets`

### Changed

* update `blank`
* update `docs-snippets`

### Fixed

* insufficient patterns for skipping references

## [0.2] -- 2022-05-10

### Added

* submodule `blank`

### Changed

* optimise `.gitignore`
* optimise repository manual build instruction
* update `docs-snippets`

### Fixed

* add missing `catch`
* docstrings which are not cited
* obsolete line in `CITATION.cff`

## [0.1] -- 2022-05-09

### Added

* add GPL 2.0 license
* build instruction:  GHC and Haddock compilation
* build instruction:  installation
* build instruction:  repository manual
* create `.gitignore`
* create `Main.hs`
* create project meta data for repository manual
* create repository README
* create this changelog
* create this repository
* submodule `docs-snippets`
* submodule `routines`

<!----------------------------------------------------------------------------->


tests/const.rs
/*********************** GNU General Public License 2.0 ***********************\
|                                                                              |
|  Copyright (C) 2022 Kevin Matthes                                            |
|                                                                              |
|  This program is free software; you can redistribute it and/or modify        |
|  it under the terms of the GNU General Public License as published by        |
|  the Free Software Foundation; either version 2 of the License, or           |
|  (at your option) any later version.                                         |
|                                                                              |
|  This program is distributed in the hope that it will be useful,             |
|  but WITHOUT ANY WARRANTY; without even the implied warranty of              |
|  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               |
|  GNU General Public License for more details.                                |
|                                                                              |
|  You should have received a copy of the GNU General Public License along     |
|  with this program; if not, write to the Free Software Foundation, Inc.,     |
|  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.                 |
|                                                                              |
\******************************************************************************/

//! Tests for the configured constants.
use cffreference::EX_USAGE;

/// Test the method `fix_level`.
#[test]
fn ex_usage() {
    assert_eq!(64i32, EX_USAGE)
}

/******************************************************************************/


tests/trigger.rs
/*********************** GNU General Public License 2.0 ***********************\
|                                                                              |
|  Copyright (C) 2022 Kevin Matthes                                            |
|                                                                              |
|  This program is free software; you can redistribute it and/or modify        |
|  it under the terms of the GNU General Public License as published by        |
|  the Free Software Foundation; either version 2 of the License, or           |
|  (at your option) any later version.                                         |
|                                                                              |
|  This program is distributed in the hope that it will be useful,             |
|  but WITHOUT ANY WARRANTY; without even the implied warranty of              |
|  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               |
|  GNU General Public License for more details.                                |
|                                                                              |
|  You should have received a copy of the GNU General Public License along     |
|  with this program; if not, write to the Free Software Foundation, Inc.,     |
|  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.                 |
|                                                                              |
\******************************************************************************/

//! Tests for the `Trigger` struct.
use cffreference::Trigger;

/// Test the method `help`.
#[test]
fn help() {
    assert_eq!("help", setup().help())
}

/// Test the method `input`.
#[test]
fn input() {
    assert_eq!("input", setup().input())
}

/// Test the method `license`.
#[test]
fn license() {
    assert_eq!("license", setup().license())
}

/// Test the method `output`.
#[test]
fn output() {
    assert_eq!("output", setup().output())
}

/// Create a new object to test.
fn setup() -> Trigger {
    Trigger::new("help", "input", "license", "output", "version")
}

/// Test the method `version`.
#[test]
fn version() {
    assert_eq!("version", setup().version())
}

/******************************************************************************/


tests/version.rs
/*********************** GNU General Public License 2.0 ***********************\
|                                                                              |
|  Copyright (C) 2022 Kevin Matthes                                            |
|                                                                              |
|  This program is free software; you can redistribute it and/or modify        |
|  it under the terms of the GNU General Public License as published by        |
|  the Free Software Foundation; either version 2 of the License, or           |
|  (at your option) any later version.                                         |
|                                                                              |
|  This program is distributed in the hope that it will be useful,             |
|  but WITHOUT ANY WARRANTY; without even the implied warranty of              |
|  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               |
|  GNU General Public License for more details.                                |
|                                                                              |
|  You should have received a copy of the GNU General Public License along     |
|  with this program; if not, write to the Free Software Foundation, Inc.,     |
|  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.                 |
|                                                                              |
\******************************************************************************/

//! Tests for the `Version` struct.
use cffreference::Version;

/// Test the method `fix_level`.
#[test]
fn fix_level() {
    assert_eq!(3, *setup().fix_level())
}

/// Test the method `major`.
#[test]
fn major() {
    assert_eq!(1, *setup().major())
}

/// Test the method `minor`.
#[test]
fn minor() {
    assert_eq!(2, *setup().minor())
}

/// Create a new object to test.
fn setup() -> Version {
    Version::new(1, 2, 3)
}

/******************************************************************************/


src/main.rs
/*********************** GNU General Public License 2.0 ***********************\
|                                                                              |
|  Copyright (C) 2022 Kevin Matthes                                            |
|                                                                              |
|  This program is free software; you can redistribute it and/or modify        |
|  it under the terms of the GNU General Public License as published by        |
|  the Free Software Foundation; either version 2 of the License, or           |
|  (at your option) any later version.                                         |
|                                                                              |
|  This program is distributed in the hope that it will be useful,             |
|  but WITHOUT ANY WARRANTY; without even the implied warranty of              |
|  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               |
|  GNU General Public License for more details.                                |
|                                                                              |
|  You should have received a copy of the GNU General Public License along     |
|  with this program; if not, write to the Free Software Foundation, Inc.,     |
|  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.                 |
|                                                                              |
\******************************************************************************/

/// The main function.
///
/// This main function takes care about the application's in- and output and
/// composes its overall behaviour.  The therefore required components are
/// outsourced to the library crate root of this project.
fn main() -> Result<(), Box<dyn std::error::Error>> {
    let trigger = cffreference::Trigger::default();
    let version = cffreference::Version::new(1, 0, 0);

    let mut options: getopts::Options = getopts::Options::new();
    options
        .optflag("h", trigger.help(), "Show this help and exit.")
        .optopt(
            "i",
            trigger.input(),
            "The input file to read from.",
            "<FILE>",
        )
        .optflag(
            "l",
            trigger.license(),
            "Show the license information and exit.",
        )
        .optopt(
            "o",
            trigger.output(),
            "The output file to write to.",
            "<FILE>",
        )
        .optflag(
            "v",
            trigger.version(),
            "Show the version information and exit.",
        );

    match cffreference::Application::new(
        &match cffreference::Configuration::parse(std::env::args(), &options, &trigger) {
            Ok(configuration) => configuration,
            Err(getopts::Fail::ArgumentMissing(string)) => {
                println!(
                    "{}",
                    options.usage(&format!(
                        "No argument for option '{string}' supplied.  \
                        Valid option calls are as follows:"
                    ))
                );
                std::process::exit(cffreference::EX_USAGE);
            }
            Err(getopts::Fail::UnrecognizedOption(string)) => {
                println!(
                    "{}",
                    options.usage(&format!(
                        "Unresolvable argument:  '{string}'.  \
                        Valid options are as follows:"
                    ))
                );
                std::process::exit(cffreference::EX_USAGE);
            }
            Err(error) => {
                return Err(error.into());
            }
        },
        &options,
        &version,
    )
    .run()
    {
        Ok(_) => Ok(()),
        Err(error) => Err(error),
    }
}

/******************************************************************************/


ghc-haddock.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
%% Copyright (C) 2022 Kevin Matthes
%%
%% This program is free software; you can redistribute it and/or modify
%% it under the terms of the GNU General Public License as published by
%% the Free Software Foundation; either version 2 of the License, or
%% (at your option) any later version.
%%
%% This program is distributed in the hope that it will be useful,
%% but WITHOUT ANY WARRANTY; without even the implied warranty of
%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%% GNU General Public License for more details.
%%
%% You should have received a copy of the GNU General Public License along
%% with this program; if not, write to the Free Software Foundation, Inc.,
%% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
%%
%%%%
%%
%%  FILE
%%      ghc-haddock.m
%%
%%  BRIEF
%%      Compile a Haskell project together with its documentation.
%%
%%  AUTHOR
%%      Kevin Matthes
%%
%%  COPYRIGHT
%%      (C) 2022 Kevin Matthes.
%%      This file is licensed GPL 2 as of June 1991.
%%
%%  DATE
%%      2022
%%
%%  NOTE
%%      See `LICENSE' for full license.
%%      See `README.md' for project details.
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%
%%
%% Variables.
%%
%%%%

% Software.
compiler.args   = '--make -static -fasm -fPIE -O2';
compiler.hidir  = ['-hidir' ' ' 'build'];
compiler.in     = '*.hs';
compiler.odir   = ['-odir' ' ' 'build'];
compiler.out    = ['./' 'cffreference'];
compiler.self   = 'ghc';
compiler.call   = [ compiler.self ' ' compiler.args ' ' compiler.hidir ' '   ...
                    compiler.odir ' ' compiler.in ' -o ' compiler.out        ...
                  ];

haddock.args    = '--html';
haddock.in      = '*.hs';
haddock.out     = ['./' 'html/'];
haddock.self    = 'haddock';
haddock.call    = [ haddock.self ' ' haddock.args ' -o ' haddock.out ' '     ...
                    haddock.in                                               ...
                  ];



% Miscellaneous.
misc.self   = 'ghc-haddock.m';
misc.banner = ['[ ' misc.self ' ] '];



%%%%
%%
%% Build steps.
%%
%%%%

% Begin build instruction.
disp ([misc.banner 'Begin build instruction.']);

% Clean outdated compilation.
if length (glob (compiler.out));
    fprintf ([misc.banner 'Remove outdated compilation ... ']);
    delete (compiler.out);
    disp (compiler.out);
end;

% Compile Haskell code with GHC.
disp ([misc.banner 'Compile ' compiler.out ' with GHC ...']);
disp (compiler.call);
system (compiler.call);
disp ([misc.banner 'Done.']);

% Compile Haddock documentation, if possible.
if length (glob (compiler.out));
    disp ([misc.banner 'Compile Haddock documentation ...']);
    disp (haddock.call);
    system (haddock.call);
    disp ([misc.banner 'Done.']);
end;

% End build instruction.
disp ([misc.banner 'End build instruction.']);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


src/lib.rs
/*********************** GNU General Public License 2.0 ***********************\
|                                                                              |
|  Copyright (C) 2022 Kevin Matthes                                            |
|                                                                              |
|  This program is free software; you can redistribute it and/or modify        |
|  it under the terms of the GNU General Public License as published by        |
|  the Free Software Foundation; either version 2 of the License, or           |
|  (at your option) any later version.                                         |
|                                                                              |
|  This program is distributed in the hope that it will be useful,             |
|  but WITHOUT ANY WARRANTY; without even the implied warranty of              |
|  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the               |
|  GNU General Public License for more details.                                |
|                                                                              |
|  You should have received a copy of the GNU General Public License along     |
|  with this program; if not, write to the Free Software Foundation, Inc.,     |
|  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.                 |
|                                                                              |
\******************************************************************************/

//! The crate root.
//!
//! This source file defines the data structures, magic numbers as well as
//! utility functions of this project.

/// The passed command line options are invalid.
pub const EX_USAGE: i32 = 64;

/// The application's behaviour.
pub struct Application {
    /// This application's `configuration`.
    configuration: Configuration,

    /// Information about this application.
    description: Description,

    /// The input file to read from.
    input: Option<String>,

    /// The output file to write to.
    output: Option<String>,

    /// The version information.
    version: Version,
}

impl Application {
    /// Read the input.
    fn preprocess(&self) -> Result<String, std::io::Error> {
        match std::fs::read_to_string(self.input.as_ref().unwrap()) {
            Ok(string) => Ok(string
                .lines()
                .filter(|x| {
                    !x.is_empty()
                        && !x.starts_with('#')
                        && !x.starts_with("---")
                        && !x.starts_with("```")
                        && !x.starts_with("cff-version:")
                        && !x.starts_with("message:")
                })
                .map(|x| String::from(x) + "\n")
                .collect()),
            Err(error) => match error.kind() {
                std::io::ErrorKind::NotFound => {
                    eprintln!("File '{}' not found.", self.input.as_ref().unwrap());
                    std::process::exit(EX_USAGE);
                }
                _ => Err(error),
            },
        }
    }

    /// Show the license information.
    fn license(&self) {
        println!(
            "\
            Copyright (C) 2022 Kevin Matthes\n\
            \n\
            This program is free software; you can redistribute it and/or modify\n\
            it under the terms of the GNU General Public License as published by\n\
            the Free Software Foundation; either version 2 of the License, or\n\
            (at your option) any later version.\n\
            \n\
            This program is distributed in the hope that it will be useful,\n\
            but WITHOUT ANY WARRANTY; without even the implied warranty of\n\
            MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n\
            GNU General Public License for more details.\n\
            \n\
            You should have received a copy of the GNU General Public License along\n\
            with this program; if not, write to the Free Software Foundation, Inc.,\n\
            51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.\
            "
        );
    }

    /// Create a new application with the configured behaviour.
    pub fn new(
        configuration: &Configuration,
        options: &getopts::Options,
        version: &Version,
    ) -> Self {
        let name = "cffreference";
        let synopsis = "[OPTIONS]";
        let usage = options.usage(&format!("Usage:  {name} {synopsis}"));

        Self {
            configuration: configuration.clone(),
            description: Description::new(
                name,
                "A simple CLI to extract citation information from a CITATION.cff.",
                synopsis,
                &usage,
            ),
            input: configuration
                .matches()
                .opt_str(configuration.trigger().input()),
            output: configuration
                .matches()
                .opt_str(configuration.trigger().output()),
            version: version.clone(),
        }
    }

    /// Run the application with the configured behaviour.
    pub fn run(&self) -> Result<(), Box<dyn std::error::Error>> {
        if self
            .configuration
            .matches()
            .opt_present(self.configuration.trigger().version())
        {
            self.version();
        } else if self
            .configuration
            .matches()
            .opt_present(self.configuration.trigger().license())
        {
            self.license();
        } else if self
            .configuration
            .matches()
            .opt_present(self.configuration.trigger().input())
        {
            println!(
                "Input:  {}\nOutput:  {}",
                self.input
                    .as_ref()
                    .ok_or_else(|| getopts::Fail::ArgumentMissing(String::from(
                        self.configuration.trigger().input()
                    )))?,
                if self
                    .configuration
                    .matches()
                    .opt_present(self.configuration.trigger().output())
                {
                    self.output.as_ref().ok_or_else(|| {
                        getopts::Fail::ArgumentMissing(String::from(
                            self.configuration.trigger().output(),
                        ))
                    })?
                } else {
                    "stdout"
                }
            );
            print!("{}", self.preprocess()?);
        } else {
            self.usage();
        }

        Ok(())
    }

    /// A brief in-app documentation.
    ///
    /// This function will write a brief usage information, including a short
    /// introduction to the meaning of the configured `options`, to `stdout`.
    fn usage(&self) {
        println!(
            "{}, version {}.{}.{}.\n{}\n\n{}",
            self.description.name(),
            self.version.major(),
            self.version.minor(),
            self.version.fix_level(),
            self.description.summary(),
            self.description.usage()
        );
    }

    /// A brief version information function.
    ///
    /// This function will write a short version information to `stdout`.
    fn version(&self) {
        println!(
            "{}.{}.{}",
            self.version.major(),
            self.version.minor(),
            self.version.fix_level()
        );
    }
}

/// This application's configuration.
///
/// When calling this application with command line options, its environment and
/// resulting behaviour will be set.  This struct controls which functionality
/// will be invoked.
#[derive(Clone)]
pub struct Configuration {
    /// The found options.
    matches: getopts::Matches,

    /// The configured triggers for the implemented modes.
    trigger: Trigger,
}

impl Configuration {
    /// The getter for the `matches` field.
    ///
    /// This function returns a read-only reference to the `matches` of the
    /// configured the options against the command line arguments.
    pub fn matches(&self) -> &getopts::Matches {
        &self.matches
    }

    /// Parse a new configuration from the command line arguments.
    ///
    /// In case that parsing should fail due to some reasons, the according
    /// error result will be returned immediately.
    pub fn parse(
        args: std::env::Args,
        options: &getopts::Options,
        trigger: &Trigger,
    ) -> Result<Self, getopts::Fail> {
        let matches = match options.parse(args) {
            Ok(matches) => matches,
            Err(error) => {
                return Err(error);
            }
        };

        Ok(Self {
            matches,
            trigger: trigger.clone(),
        })
    }

    /// The getter for the `trigger` field.
    pub fn trigger(&self) -> &Trigger {
        &self.trigger
    }
}

/// The application's self description.
pub struct Description {
    /// The application's `name`.
    name: String,

    /// A brief idea what this application actually does.
    summary: String,

    /// The summary of the mandatory arguments to pass.
    synopsis: String,

    /// The description of the configured options.
    usage: String,
}

impl Description {
    /// Retrieve the application's `name`.
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Create a new description from the given information.
    pub fn new(name: &str, summary: &str, synopsis: &str, usage: &str) -> Self {
        Self {
            name: String::from(name),
            summary: String::from(summary),
            synopsis: String::from(synopsis),
            usage: String::from(usage),
        }
    }

    /// Retrieve the application's `summary`.
    pub fn summary(&self) -> &str {
        &self.summary
    }

    /// Retrieve the application's `synopsis`.
    pub fn synopsis(&self) -> &str {
        &self.synopsis
    }

    /// Retrieve the application's `usage` information.
    pub fn usage(&self) -> &str {
        &self.usage
    }
}

/// The option's names which trigger a certain mode.
#[derive(Clone, Debug)]
pub struct Trigger {
    /// The long option for the `help` mode.
    help: String,

    /// The `input` file to read from.
    input: String,

    /// The long option for the `license` information mode.
    license: String,

    /// The `output` file to write to.
    output: String,

    /// The long option for the `version` information mode.
    version: String,
}

impl Trigger {
    /// Retrieve the `help` mode trigger.
    pub fn help(&self) -> &str {
        &self.help
    }

    /// Retrieve the `input` file's name.
    pub fn input(&self) -> &str {
        &self.input
    }

    /// Retrieve the `license` mode trigger.
    pub fn license(&self) -> &str {
        &self.license
    }

    /// Create a new set of trigger options.
    pub fn new(help: &str, input: &str, license: &str, output: &str, version: &str) -> Self {
        Self {
            help: String::from(help),
            input: String::from(input),
            license: String::from(license),
            output: String::from(output),
            version: String::from(version),
        }
    }

    /// Retrieve the `output` file's name.
    pub fn output(&self) -> &str {
        &self.output
    }

    /// Retrieve the `version` mode trigger.
    pub fn version(&self) -> &str {
        &self.version
    }
}

impl Default for Trigger {
    fn default() -> Trigger {
        Self {
            help: String::from("help"),
            input: String::from("input"),
            license: String::from("license"),
            output: String::from("output"),
            version: String::from("version"),
        }
    }
}

impl PartialEq for Trigger {
    fn eq(&self, other: &Trigger) -> bool {
        self.help() == other.help()
            && self.input() == other.input()
            && self.license() == other.license()
            && self.output() == other.output()
            && self.version() == other.version()
    }
}

/// The version information.
#[derive(Clone)]
pub struct Version {
    /// The application's major version.
    major: u32,

    /// The application's minor version.
    minor: u32,

    /// The application's fix level.
    fix_level: u32,
}

impl Version {
    /// Retrieve the fix level.
    pub fn fix_level(&self) -> &u32 {
        &self.fix_level
    }

    /// Retrieve the major version.
    pub fn major(&self) -> &u32 {
        &self.major
    }

    /// Retrieve the minor version.
    pub fn minor(&self) -> &u32 {
        &self.minor
    }

    /// Construct a new version information.
    pub fn new(major: u32, minor: u32, fix_level: u32) -> Self {
        Self {
            major,
            minor,
            fix_level,
        }
    }
}

/******************************************************************************/
